### PR TITLE
Rewrote the Audio Decoder

### PR DESCRIPTION
**Pull request by DrChat**
* Can build libav ourselves now
* Decodes frame-by-frame, so no more weirdness if the game specifies a specific offset.
 * Also supports audio loops
* If a frame fails to decode, will skip to the next frame.
* Crapload of error-checking to make sure everything is running smoothly
* Code is now 100% more spaghettified!

* Should fix audio dropping out due to decoder errors. Unfortunately, there's a deadlock in XAudio2 which will still cause audio to die in some games.

### CODE REVIEW

File: src/xenia/app/premake5.lua [hunk 1]
```diff
@@ -32,6 +32,7 @@ project("xenia-app")
   })
   includedirs({
     project_root.."/third_party/elemental-forms/src",
+	project_root.."/build_tools/third_party/gflags/src",
```
Comment (benvanik) [inline src/xenia/app/premake5.lua]: tab

File: src/xenia/app/premake5.lua [hunk 2]
```diff
@@ -52,6 +53,8 @@ project("xenia-app")
       "xenia-hid-winkey",
       "xenia-hid-xinput",
     })
+
+--[[
```
Comment (benvanik) [inline src/xenia/app/premake5.lua]: No longer needed - can kill this

File: src/xenia/apu/nop/premake5.lua [hunk 3]
```diff
@@ -13,5 +13,6 @@ project("xenia-apu-nop")
   defines({
   })
   includedirs({
+	project_root.."/build_tools/third_party/gflags/src",
```
Comment (benvanik) [inline src/xenia/apu/nop/premake5.lua]: tab (here and in other lua files)

File: src/xenia/apu/xma_context.cc [hunk 4]
```diff
@@ -243,162 +346,319 @@ void XmaContext::DecodePackets(XMA_CONTEXT_DATA* data) {
...
-    uint32_t input_offset_bytes = seq_offset_bytes;
-    auto input_buffer = in0;
-
-    if (seq_offset_bytes >= input_size_0_bytes && input_size_1_bytes) {
-      // Size overlap, select input buffer 1.
-      // TODO(DrChat): This needs testing.
-      input_offset_bytes -= input_size_0_bytes;
-      input_buffer = in1;
-    }
+#ifdef DEBUG
```
Comment (benvanik) [inline src/xenia/apu/xma_context.cc]: Doesn't seem too expensive - would be good to keep it always in. Avoiding behavioral changes in debug vs. release builds is really nice.

File: src/xenia/apu/xma_decoder.cc [hunk 5]
```diff
@@ -55,7 +55,34 @@ XmaDecoder::XmaDecoder(cpu::Processor* processor)
 XmaDecoder::~XmaDecoder() = default;
 
 void av_log_callback(void* avcl, int level, const char* fmt, va_list va) {
-  xe::LogLineVarargs('A', fmt, va);
+#ifdef NDEBUG
```
Comment (benvanik) [inline src/xenia/apu/xma_decoder.cc]: Would be nice to have this as a flag instead (like --verbose_libav_logging or something).

File: src/xenia/apu/xma_helpers.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
+/**
+******************************************************************************
```
Comment (benvanik) [inline src/xenia/apu/xma_helpers.h]: space

File: src/xenia/apu/xma_context.cc (hunk 4 above)
Comment (DrChat) [inline src/xenia/apu/xma_context.cc]: In practice these variables never change, so no point really in having the checks.

Comment (benvanik) [inline src/xenia/apu/xma_context.cc]: If it's useful as an assert you could just change them to that (like the buffer size check?) - otherwise kill em :)

**Pull request merged**